### ISSUE TITLE
Salt States and Orchestration output different between success and failure.

### ISSUE THREAD

**Issue by UtahDave** [saltstack/salt#44451]
### Description of Issue/Question
When a Salt State or Orchestration state are successful they output the result in the `changes` field as a dictionary.

When they fail the `changes` directory is empty and the failed results are added to the `comments` field as a string that includes hard returns and spaces. This makes it difficult to parse reliably when used programatically.

### Setup

I have a Vagrant environment that will allow this to be reproduced easily.

https://github.com/UtahDave/changes

### Steps to Reproduce Issue
Here's the output from a successful run:

```
vagrant up
vagrant ssh master
sudo salt minion1 pkg.remove nano
sudo salt-run state.orch changes.succeed --out json
{
    "outputter": "highstate",
    "data": {
        "saltmaster.local_master": {
            "salt_|-Step01_|-Step01_|-state": {
                "comment": "States ran successfully. Updating minion1.",
                "name": "Step01",
                "start_time": "01:47:59.026329",
                "result": true,
                "duration": 8006.557,
                "__run_num__": 0,
                "__jid__": "20171109014759087105",
                "__sls__": "changes.succeed",
                "changes": {
                    "ret": {
                        "minion1": {
                            "pkg_|-install_good_package_|-nano_|-installed": {
                                "comment": "The following packages were installed/updated: nano",
                                "name": "nano",
                                "start_time": "01:47:59.077133",
                                "result": true,
                                "duration": 7202.901,
                                "__run_num__": 0,
                                "__sls__": "changes.pkg_succeed",
                                "changes": {
                                    "nano": {
                                        "new": "2.5.3-2ubuntu2",
                                        "old": ""
                                    }
                                },
                                "__id__": "install_good_package"
                            }
                        }
                    },
                    "out": "highstate"
                },
                "__id__": "Step01"
            }
        }
    },
    "retcode": 0
}
```

Here's the stdout output from a failure:

```
sudo salt-run state.orch changes.fail --out json 2> /dev/null
{
    "outputter": "highstate",
    "data": {
        "saltmaster.local_master": {
            "salt_|-Step01_|-Step01_|-state": {
                "comment": "Run failed on minions: minion1\nFailures:\n    {\n        \"minion1\": {\n            \"pkg_|-install_fake_package_|-asdfasdf_|-installed\": {\n                \"comment\": \"Problem encountered installing package(s). Additional info follows:\\n\\nerrors:\\n    - Running scope as unit run-r9c32e46083d64e3785334798d98071e9.scope.\\n      E: Unable to locate package asdfasdf\", \n                \"name\": \"asdfasdf\", \n                \"start_time\": \"01:41:40.142147\", \n                \"result\": false, \n                \"duration\": 4373.974, \n                \"__run_num__\": 0, \n                \"__sls__\": \"changes.pkg_fail\", \n                \"changes\": {}, \n                \"__id__\": \"install_fake_package\"\n            }\n        }\n    }\n",
                "name": "Step01",
                "start_time": "01:41:40.036106",
                "result": false,
                "duration": 5130.725,
                "__run_num__": 0,
                "__jid__": "20171109014140096711",
                "__sls__": "changes.fail",
                "changes": {},
                "__id__": "Step01"
            }
        }
    },
    "retcode": 1
}
```

### Versions Report
```
vagrant@saltmaster:~$ sudo salt \* test.versions_report
minion1:
    Salt Version:
               Salt: 2017.7.2
     
    Dependency Versions:
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: 2.4.2
          docker-py: Not Installed
              gitdb: Not Installed
          gitpython: Not Installed
              ioflo: Not Installed
             Jinja2: 2.8
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: 1.0.3
       msgpack-pure: Not Installed
     msgpack-python: 0.4.6
       mysql-python: Not Installed
          pycparser: Not Installed
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.12 (default, Nov 19 2016, 06:48:10)
       python-gnupg: Not Installed
             PyYAML: 3.11
              PyZMQ: 15.2.0
               RAET: Not Installed
              smmap: Not Installed
            timelib: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.1.4
     
    System Versions:
               dist: Ubuntu 16.04 xenial
             locale: UTF-8
            machine: x86_64
            release: 4.4.0-87-generic
             system: Linux
            version: Ubuntu 16.04 xenial
```

The master:

```
vagrant@saltmaster:~$ sudo salt --versions-report
Salt Version:
           Salt: 2017.7.2
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Nov 19 2016, 06:48:10)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: Ubuntu 16.04 xenial
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-87-generic
         system: Linux
        version: Ubuntu 16.04 xenial
```

**Comment by UtahDave** [comment]
ZD 1847

**Comment by Ch3LL** [comment]
@UtahDave can you clarify the request? Are you expecting the comments field to not be a string but a different type? Or are you expecting the comments field to be somewhere else?

**Comment by UtahDave** [comment]
Well, the idea is that just like the `changes` field contains the info in a clean data structure when the command is successful, the error info should be made available in a clean data structure when the command fails.

The `comments` field is designed to be free form text, so that might not be the best place. Maybe we add an `errors` field?  The original customer doesn't have a strong opinion on the exact location, other than the desire to be able to have the info in a data structure that is easily accessed programmatically.

Does that help?

**Comment by Ch3LL** [comment]
ping @saltstack/team-core any ideas here as to how we could improve the comments field to be better structured or change the field possibly to be `errors` and structured?

**Comment by Ch3LL** [comment]
ping @terminalmage mind taking a look at this for 2017.7.3?

**Comment by terminalmage** [comment]
This looks to be the same issue as https://github.com/saltstack/salt/issues/43204.

**Comment by rallytime** [comment]
@UtahDave confirmed that this was a duplicate with @terminalmage. The fix for this is in #44751 and will be available in the `Oxygen` release. Closing. :)
